marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.hub/lib/flow-lib.xqy

```diff
@@ -614,15 +624,88 @@ declare function flow:run-main(
       debug:log(xdmp:describe($ex, (), ())),
 
       (: log the trace event for main :)
-      trace:set-plugin-label("main"),
-      trace:error-trace($ex, xdmp:elapsed-time() - $before)
+      trace:set-plugin-label(rfc:get-trace($item-context), "main"),
+      trace:error-trace($item-context, $ex, xdmp:elapsed-time() - $before)
     ),
     xdmp:rethrow()
   }
   return
     $resp
 };
 
+declare function flow:queue-writer(
+  $writer-function,
+  $identifier as xs:string,
+  $envelope as item(),
+  $options as map:map)
+{
+  $writer-queue =>
+    map:with($identifier,
+      map:map()
+        => map:with("writer-function", $writer-function)
+        => map:with("envelope", $envelope)
+        => map:with("options", $options)
+    )
+};
+
+declare function flow:run-writers(
+  $identifiers as xs:string*)
+{
+  let $updated-settings := xdmp:eval('
```

Comment:
We run the flows in query only mode. If you use an invoke-function then ML will think that the entire parent transaction is update mode. why you might ask? because ML uses static analysis of the call tree (or whatever CS terms I'm abusing) to see if there are any update functions being called anywhere. It's pretty lazy about it and doesn't account for the fact that an update call is within an invoke-function. and thus it wrongly marks the whole thing as update :(

Reply:
makes sense